raspberry-pi/4/audio.nix

```diff
@@ -0,0 +1,46 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.hardware.raspberry-pi."4".audio;
+in
+{
+  options.hardware = {
+    raspberry-pi."4".audio = {
+      enable = lib.mkEnableOption ''
+        Enable configuration for audio
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    hardware.deviceTree = {
+      overlays = [
+        # Equivalent to dtparam=audio=on
```

Comment:
Has the device tree any advantage over kernel parameter?

Reply:
*It works*.

That is:

 - It works in the NixOS generations lifecycle; booting a generation with the setting turned on makes it work, a generation with the setting turned off makes it work.
 - U-Boot clobbers the firmware-provided FDT (FDT is the Flattened Device Tree in memory) with the dtb file from the generation, so *pretty much all setup done by the firmware is lost*.

Reply:
OH, and I missed the most obvious part here: `dtparam=` is **not** a kernel parameter!

These are `config.txt` option that should be seen as turning bios options on/off on the Raspberry Pi. The proprietary vendor firmware consumes `dtparam` options and modify the built-in FDT.